File: articles/design-thinking-complex-institutions/julia/institutional_readiness_model.jl
#!/usr/bin/env julia

# Institutional readiness model in Julia.
# Uses only Julia standard-library functionality.

using Printf

article_dir = normpath(joinpath(@__DIR__, ".."))
input_path = joinpath(article_dir, "data", "raw", "institutional_design_options_raw.csv")
output_dir = joinpath(article_dir, "outputs")
mkpath(output_dir)

struct InstitutionalOption
    name::String
    desirability::Float64
    authority::Float64
    capability::Float64
    funding::Float64
    policy_fit::Float64
    governance::Float64
    trust_gain::Float64
    burden_reduction::Float64
    coordination::Float64
    implementation_risk::Float64
    data_readiness::Float64
    frontline_fit::Float64
    maintenance::Float64
    equity::Float64
    public_value::Float64
end

function parse_float(value)
    return parse(Float64, strip(value))
end

function load_options(path::String)
    lines = readlines(path)
    items = InstitutionalOption[]

    for line in lines[2:end]
        fields = split(line, ",")
        if length(fields) < 17
            error("Malformed option row: $(line)")
        end

        push!(
            items,
            InstitutionalOption(
                strip(fields[1]),
                parse_float(fields[3]),
                parse_float(fields[4]),
                parse_float(fields[5]),
                parse_float(fields[6]),
                parse_float(fields[7]),
                parse_float(fields[8]),
                parse_float(fields[9]),
                parse_float(fields[10]),
                parse_float(fields[11]),
                parse_float(fields[12]),
                parse_float(fields[13]),
                parse_float(fields[14]),
                parse_float(fields[15]),
                parse_float(fields[16]),
                parse_float(fields[17])
            )
        )
    end

    return items
end

function readiness(x::InstitutionalOption)
    return 0.14*x.desirability + 0.13*x.authority + 0.12*x.capability +
           0.10*x.funding + 0.10*x.policy_fit + 0.11*x.governance +
           0.08*x.trust_gain + 0.08*x.burden_reduction + 0.06*x.data_readiness +
           0.05*x.frontline_fit + 0.03*x.maintenance -
           0.05*x.coordination - 0.05*x.implementation_risk
end

function absorption(x::InstitutionalOption)
    return 0.18*x.authority + 0.18*x.capability + 0.14*x.funding +
           0.14*x.governance + 0.12*x.policy_fit + 0.10*x.frontline_fit +
           0.08*x.maintenance + 0.06*x.data_readiness
end

function public_value_priority(x::InstitutionalOption)
    return 0.24*x.public_value + 0.20*x.burden_reduction + 0.18*x.trust_gain +
           0.16*x.equity + 0.12*x.desirability + 0.10*x.policy_fit -
           0.08*x.implementation_risk
end

function sequencing_need(x::InstitutionalOption)
    return 0.28*x.coordination + 0.24*x.implementation_risk +
           0.14*(10.0-x.authority) + 0.12*(10.0-x.capability) +
           0.10*(10.0-x.funding) + 0.07*(10.0-x.maintenance) +
           0.05*(10.0-x.data_readiness)
end

function portfolio_score(x::InstitutionalOption)
    return 0.30*readiness(x) + 0.26*public_value_priority(x) +
           0.20*absorption(x) + 0.10*x.equity + 0.06*x.trust_gain -
           0.12*sequencing_need(x) - 0.08*x.implementation_risk
end

options = load_options(input_path)
scored = [(x.name, readiness(x), absorption(x), public_value_priority(x), sequencing_need(x), portfolio_score(x)) for x in options]
scored = sort(scored, by = x -> x[6], rev = true)

open(joinpath(output_dir, "julia_institutional_readiness_scores.csv"), "w") do io
    println(io, "rank,option,change_readiness,absorption_capacity,public_value_priority,sequencing_need,portfolio_score")
    for (rank, row) in enumerate(scored)
        @printf(io, "%d,%s,%.6f,%.6f,%.6f,%.6f,%.6f\n", rank, row[1], row[2], row[3], row[4], row[5], row[6])
    end
end

println("Julia institutional readiness model complete.")
println("Outputs written to: $(output_dir)")

File: articles/design-thinking-complex-institutions/fortran/institutional_readiness_model.f90
program institutional_readiness_model
  implicit none

  integer, parameter :: n = 10
  character(len=180) :: names(n)
  real(8) :: desirability(n), authority(n), capability(n), funding(n), policy_fit(n)
  real(8) :: governance(n), trust_gain(n), burden_reduction(n), coordination(n), risk(n)
  real(8) :: data_readiness(n), frontline_fit(n), maintenance(n), equity(n), public_value(n)
  real(8) :: readiness(n), absorption(n), public_priority(n), sequencing(n), portfolio(n)
  integer :: i

  names = [ character(len=180) :: &
    "Simplify eligibility pathway", &
    "Create assisted access model", &
    "Redesign cross-department handoffs", &
    "Build institutional learning dashboard", &
    "Prototype community accountability board", &
    "Modernize legacy case-management data", &
    "Redesign frontline escalation process", &
    "Create AI-assisted research synthesis workflow", &
    "Develop burden audit and repair pathway", &
    "Create policy interpretation review board" ]

  desirability = [8.8d0,9.0d0,8.2d0,7.6d0,8.4d0,7.8d0,8.6d0,7.4d0,8.8d0,7.8d0]
  authority = [6.4d0,6.8d0,5.8d0,7.2d0,5.4d0,6.0d0,7.0d0,6.6d0,6.2d0,5.8d0]
  capability = [6.8d0,6.4d0,5.6d0,7.0d0,5.8d0,5.2d0,7.2d0,6.2d0,6.4d0,5.8d0]
  funding = [6.6d0,6.2d0,5.8d0,6.8d0,5.4d0,5.0d0,6.6d0,6.0d0,6.0d0,5.4d0]
  policy_fit = [6.0d0,6.4d0,6.8d0,7.4d0,5.8d0,7.0d0,7.2d0,6.6d0,6.6d0,7.6d0]
  governance = [6.2d0,6.4d0,5.6d0,7.2d0,5.8d0,6.0d0,7.0d0,6.4d0,6.8d0,6.4d0]
  trust_gain = [8.2d0,8.8d0,7.6d0,6.8d0,9.0d0,6.4d0,7.8d0,6.6d0,8.6d0,7.4d0]
  burden_reduction = [8.6d0,8.4d0,7.4d0,6.2d0,7.8d0,6.8d0,8.0d0,6.4d0,9.0d0,7.6d0]
  coordination = [6.8d0,7.2d0,8.6d0,6.4d0,7.8d0,8.8d0,6.6d0,7.4d0,7.2d0,7.8d0]
  risk = [5.8d0,6.2d0,7.4d0,5.6d0,6.8d0,8.0d0,5.4d0,6.6d0,6.0d0,6.8d0]
  data_readiness = [6.0d0,5.8d0,6.2d0,7.4d0,5.6d0,5.4d0,6.8d0,7.0d0,6.2d0,6.0d0]
  frontline_fit = [6.8d0,7.0d0,6.0d0,6.6d0,6.0d0,5.4d0,8.0d0,6.2d0,7.0d0,6.2d0]
  maintenance = [6.2d0,6.4d0,5.8d0,6.8d0,5.6d0,5.0d0,7.0d0,6.4d0,6.6d0,5.8d0]
  equity = [8.4d0,9.0d0,7.6d0,6.8d0,9.2d0,7.2d0,7.8d0,6.8d0,9.0d0,8.0d0]
  public_value = [8.8d0,9.0d0,8.0d0,7.4d0,8.6d0,7.6d0,8.4d0,7.2d0,8.8d0,8.2d0]

  do i = 1, n
    readiness(i) = 0.14d0*desirability(i) + 0.13d0*authority(i) + &
                   0.12d0*capability(i) + 0.10d0*funding(i) + &
                   0.10d0*policy_fit(i) + 0.11d0*governance(i) + &
                   0.08d0*trust_gain(i) + 0.08d0*burden_reduction(i) + &
                   0.06d0*data_readiness(i) + 0.05d0*frontline_fit(i) + &
                   0.03d0*maintenance(i) - 0.05d0*coordination(i) - 0.05d0*risk(i)

    absorption(i) = 0.18d0*authority(i) + 0.18d0*capability(i) + &
                    0.14d0*funding(i) + 0.14d0*governance(i) + &
                    0.12d0*policy_fit(i) + 0.10d0*frontline_fit(i) + &
                    0.08d0*maintenance(i) + 0.06d0*data_readiness(i)

    public_priority(i) = 0.24d0*public_value(i) + 0.20d0*burden_reduction(i) + &
                         0.18d0*trust_gain(i) + 0.16d0*equity(i) + &
                         0.12d0*desirability(i) + 0.10d0*policy_fit(i) - &
                         0.08d0*risk(i)

    sequencing(i) = 0.28d0*coordination(i) + 0.24d0*risk(i) + &
                    0.14d0*(10.0d0-authority(i)) + 0.12d0*(10.0d0-capability(i)) + &
                    0.10d0*(10.0d0-funding(i)) + 0.07d0*(10.0d0-maintenance(i)) + &
                    0.05d0*(10.0d0-data_readiness(i))

    portfolio(i) = 0.30d0*readiness(i) + 0.26d0*public_priority(i) + &
                   0.20d0*absorption(i) + 0.10d0*equity(i) + &
                   0.06d0*trust_gain(i) - 0.12d0*sequencing(i) - 0.08d0*risk(i)
  end do

  call sort_desc(names, readiness, absorption, public_priority, sequencing, portfolio, n)

  print '(a)', 'rank,option,change_readiness,absorption_capacity,public_value_priority,sequencing_need,portfolio_score'
  do i = 1, n
    print '(i0,a,a,a,f7.4,a,f7.4,a,f7.4,a,f7.4,a,f7.4)', &
      i, ',', trim(names(i)), ',', readiness(i), ',', absorption(i), ',', public_priority(i), ',', sequencing(i), ',', portfolio(i)
  end do

contains

  subroutine sort_desc(names, readiness, absorption, public_priority, sequencing, portfolio, n)
    integer, intent(in) :: n
    character(len=180), intent(inout) :: names(n)
    real(8), intent(inout) :: readiness(n), absorption(n), public_priority(n), sequencing(n), portfolio(n)
    integer :: i, j
    real(8) :: tr, ta, tp, ts, tpo
    character(len=180) :: tn

    do i = 1, n - 1
      do j = i + 1, n
        if (portfolio(j) > portfolio(i)) then
          tpo = portfolio(i); portfolio(i) = portfolio(j); portfolio(j) = tpo
          tr = readiness(i); readiness(i) = readiness(j); readiness(j) = tr
          ta = absorption(i); absorption(i) = absorption(j); absorption(j) = ta
          tp = public_priority(i); public_priority(i) = public_priority(j); public_priority(j) = tp
          ts = sequencing(i); sequencing(i) = sequencing(j); sequencing(j) = ts
          tn = names(i); names(i) = names(j); names(j) = tn
        end if
      end do
    end do
  end subroutine sort_desc

end program institutional_readiness_model
